File: app/src/main/java/com/bookmyspace/bookmyspace/ui/components/ResponsiveLayout.kt
package com.bookmyspace.bookmyspace.ui.components

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyListScope
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

/**
 * Material 3 Window Width Size Classes for adaptive layouts
 */
enum class ResponsiveWindowWidthClass {
    COMPACT,   // < 600dp (standard portrait phones)
    MEDIUM,    // 600dp - 839dp (foldables unfolded, small tablets, landscape phones)
    EXPANDED   // >= 840dp (large tablets, desktop, ChromeOS)
}

/**
 * Data class holding responsive sizing metrics and layout configurations
 */
data class ResponsiveDimensions(
    val windowWidthClass: ResponsiveWindowWidthClass,
    val availableWidth: Dp,
    val isCompact: Boolean,
    val isMedium: Boolean,
    val isExpanded: Boolean,
    val isTabletOrWide: Boolean,
    val categoryGridColumns: Int,
    val resultsGridColumns: Int,
    val horizontalPadding: Dp,
    val gridSpacing: Dp,
    val maxContentWidth: Dp
)

/**
 * Calculates ResponsiveDimensions based on current available width
 */
fun calculateResponsiveDimensions(width: Dp): ResponsiveDimensions {
    val windowClass = when {
        width < 600.dp -> ResponsiveWindowWidthClass.COMPACT
        width < 840.dp -> ResponsiveWindowWidthClass.MEDIUM
        else -> ResponsiveWindowWidthClass.EXPANDED
    }

    val isCompact = windowClass == ResponsiveWindowWidthClass.COMPACT
    val isMedium = windowClass == ResponsiveWindowWidthClass.MEDIUM
    val isExpanded = windowClass == ResponsiveWindowWidthClass.EXPANDED
    val isTabletOrWide = !isCompact

    return ResponsiveDimensions(
        windowWidthClass = windowClass,
        availableWidth = width,
        isCompact = isCompact,
        isMedium = isMedium,
        isExpanded = isExpanded,
        isTabletOrWide = isTabletOrWide,
        categoryGridColumns = if (isCompact) 1 else 2,
        resultsGridColumns = if (isCompact) 1 else if (isMedium) 2 else 2,
        horizontalPadding = if (isCompact) 20.dp else if (isMedium) 28.dp else 36.dp,
        gridSpacing = if (isCompact) 12.dp else 16.dp,
        maxContentWidth = 1100.dp
    )
}

/**
 * Composable builder that measures available constraints and provides [ResponsiveDimensions]
 */
@Composable
fun ResponsiveLayout(
    modifier: Modifier = Modifier,
    content: @Composable BoxScope.(ResponsiveDimensions) -> Unit
) {
    BoxWithConstraints(
        modifier = modifier.fillMaxSize(),
        contentAlignment = Alignment.TopCenter
    ) {
        val responsiveInfo = remember(maxWidth) {
            calculateResponsiveDimensions(maxWidth)
        }

        Box(
            modifier = Modifier
                .fillMaxSize()
                .widthIn(max = responsiveInfo.maxContentWidth)
        ) {
            content(responsiveInfo)
        }
    }
}

/**
 * Extension for LazyListScope to render items in a responsive grid across rows
 */
fun <T> LazyListScope.responsiveGridItems(
    items: List<T>,
    columns: Int,
    key: ((T) -> Any)? = null,
    horizontalSpacing: Dp = 12.dp,
    verticalSpacing: Dp = 12.dp,
    contentPadding: PaddingValues = PaddingValues(horizontal = 20.dp),
    itemContent: @Composable (item: T, index: Int) -> Unit
) {
    if (columns <= 1) {
        // Single column layout
        items(
            count = items.size,
            key = if (key != null) { index -> key(items[index]) } else null
        ) { index ->
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(contentPadding)
                    .padding(vertical = verticalSpacing / 2)
            ) {
                itemContent(items[index], index)
            }
        }
    } else {
        // Multi-column grid chunked into rows
        val chunkedItems = items.chunked(columns)
        items(
            count = chunkedItems.size,
            key = if (key != null) { rowIndex ->
                chunkedItems[rowIndex].joinToString("-") { key(it).toString() }
            } else null
        ) { rowIndex ->
            val rowItems = chunkedItems[rowIndex]
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(contentPadding)
                    .padding(vertical = verticalSpacing / 2),
                horizontalArrangement = Arrangement.spacedBy(horizontalSpacing),
                verticalAlignment = Alignment.CenterVertically
            ) {
                rowItems.forEachIndexed { itemIndex, item ->
                    val globalIndex = rowIndex * columns + itemIndex
                    Box(modifier = Modifier.weight(1f)) {
                        itemContent(item, globalIndex)
                    }
                }
                // Fill remainder of row with empty spaces for balanced alignment
                val remainder = columns - rowItems.size
                if (remainder > 0) {
                    repeat(remainder) {
                        Spacer(modifier = Modifier.weight(1f))
                    }
                }
            }
        }
    }
}

File: lib/core/widgets/responsive_layout.dart
import 'package:flutter/material.dart';

/// Screen width breakpoints based on Material 3 specifications.
enum ResponsiveWindowSizeClass {
  compact,   // < 600dp (standard portrait phones)
  medium,    // 600dp - 839dp (foldables, small tablets, portrait tablets)
  expanded,  // 840dp - 1199dp (medium/large tablets, desktop)
  extraWide, // >= 1200dp (extra-wide tablet landscape, large monitors)
}

/// Sizing and layout metrics computed for the current screen constraints.
class ResponsiveInfo {
  const ResponsiveInfo({
    required this.windowClass,
    required this.availableWidth,
    required this.availableHeight,
    required this.isCompact,
    required this.isMedium,
    required this.isExpanded,
    required this.isExtraWide,
    required this.isTabletOrLandscape,
    required this.categoryColumns,
    required this.categoryAspectRatio,
    required this.resultsColumns,
    required this.resultsAspectRatio,
    required this.horizontalPadding,
    required this.gridSpacing,
    required this.maxContentWidth,
  });

  final ResponsiveWindowSizeClass windowClass;
  final double availableWidth;
  final double availableHeight;
  final bool isCompact;
  final bool isMedium;
  final bool isExpanded;
  final bool isExtraWide;
  final bool isTabletOrLandscape;
  final int categoryColumns;
  final double categoryAspectRatio;
  final int resultsColumns;
  final double resultsAspectRatio;
  final double horizontalPadding;
  final double gridSpacing;
  final double maxContentWidth;

  factory ResponsiveInfo.fromConstraints(BoxConstraints constraints) {
    final width = constraints.maxWidth;
    final height = constraints.maxHeight;

    final ResponsiveWindowSizeClass windowClass;
    if (width < 600) {
      windowClass = ResponsiveWindowSizeClass.compact;
    } else if (width < 840) {
      windowClass = ResponsiveWindowSizeClass.medium;
    } else if (width < 1200) {
      windowClass = ResponsiveWindowSizeClass.expanded;
    } else {
      windowClass = ResponsiveWindowSizeClass.extraWide;
    }

    final isCompact = windowClass == ResponsiveWindowSizeClass.compact;
    final isMedium = windowClass == ResponsiveWindowSizeClass.medium;
    final isExpanded = windowClass == ResponsiveWindowSizeClass.expanded;
    final isExtraWide = windowClass == ResponsiveWindowSizeClass.extraWide;
    final isTabletOrLandscape = !isCompact;

    // Responsive category columns and dynamic aspect ratios
    final int categoryColumns;
    final double categoryAspectRatio;
    if (isCompact) {
      categoryColumns = 1;
      categoryAspectRatio = width < 360 ? 2.3 : 2.7;
    } else if (isMedium) {
      categoryColumns = 2;
      categoryAspectRatio = 2.1;
    } else if (isExpanded) {
      categoryColumns = 2;
      categoryAspectRatio = 1.9;
    } else {
      // Extra-wide landscape: 4 columns in a single balanced horizontal hero row
      categoryColumns = 4;
      categoryAspectRatio = 1.35;
    }

    // Responsive venue/space result cards columns and aspect ratios
    final int resultsColumns;
    final double resultsAspectRatio;
    if (isCompact) {
      resultsColumns = 1;
      resultsAspectRatio = 0.88;
    } else if (isMedium) {
      resultsColumns = 2;
      resultsAspectRatio = 0.82;
    } else if (isExpanded) {
      resultsColumns = 3;
      resultsAspectRatio = 0.78;
    } else {
      resultsColumns = 4;
      resultsAspectRatio = 0.75;
    }

    final double horizontalPadding = isCompact
        ? 16.0
        : isMedium
            ? 24.0
            : isExpanded
                ? 32.0
                : 40.0;

    final double gridSpacing = isCompact ? 12.0 : 16.0;

    return ResponsiveInfo(
      windowClass: windowClass,
      availableWidth: width,
      availableHeight: height,
      isCompact: isCompact,
      isMedium: isMedium,
      isExpanded: isExpanded,
      isExtraWide: isExtraWide,
      isTabletOrLandscape: isTabletOrLandscape,
      categoryColumns: categoryColumns,
      categoryAspectRatio: categoryAspectRatio,
      resultsColumns: resultsColumns,
      resultsAspectRatio: resultsAspectRatio,
      horizontalPadding: horizontalPadding,
      gridSpacing: gridSpacing,
      maxContentWidth: 1280.0,
    );
  }
}

/// A responsive container that measures available space, calculates sizing metrics,
/// and constrains content width for optimal readability across phones, tablets, and desktop.
class ResponsiveLayoutBuilder extends StatelessWidget {
  const ResponsiveLayoutBuilder({
    super.key,
    required this.builder,
  });

  final Widget Function(BuildContext context, ResponsiveInfo responsive) builder;

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (context, constraints) {
        final responsive = ResponsiveInfo.fromConstraints(constraints);
        return Center(
          child: ConstrainedBox(
            constraints: BoxConstraints(maxWidth: responsive.maxContentWidth),
            child: builder(context, responsive),
          ),
        );
      },
    );
  }
}
